firedrake/assemble.py

```diff
@@ -169,33 +399,45 @@ def assemble_base_form(expression, tensor=None, bcs=None,
         expr = preprocess_base_form(expression, mat_type, form_compiler_parameters)
         if isinstance(expression, ufl.form.Form) and isinstance(expr, ufl.form.Form):
             expr._cache = expression._cache
+        # BaseForm preprocessing can turn BaseForm into an Expr (cf. case (6) in `restructure_base_form`)
+        if isinstance(expr, ufl.core.expr.Expr) and not isinstance(expr, ufl.core.base_form_operator.BaseFormOperator):
+            return _assemble_expr(expr)
     else:
         expr = expression
 
+    # Define assembly DAG visitor
+    assembly_visitor = functools.partial(base_form_assembly_visitor, bcs=bcs, diagonal=diagonal,
+                                         form_compiler_parameters=form_compiler_parameters,
+                                         mat_type=mat_type, sub_mat_type=sub_mat_type,
+                                         appctx=appctx, options_prefix=options_prefix,
+                                         zero_bc_nodes=zero_bc_nodes, weight=weight)
+
+    def visitor(e, *operands):
+        t = tensor if e is expr else None
+        return assembly_visitor(e, t, *operands)
+
     # DAG assembly: traverse the DAG in a post-order fashion and evaluate the node on the fly.
-    stack = [expr]
-    visited = {}
-    while stack:
-        e = stack.pop()
-        unvisted_children = []
-        operands = base_form_operands(e)
-        for arg in operands:
-            if arg not in visited:
-                unvisted_children.append(arg)
+    visited = visited or {}
+    result = base_form_postorder_traversal(expr, visitor, visited)
 
-        if unvisted_children:
-            stack.append(e)
-            stack.extend(unvisted_children)
-        else:
-            visited[e] = base_form_assembly_visitor(e, tensor, bcs, diagonal,
-                                                    form_compiler_parameters,
-                                                    mat_type, sub_mat_type,
-                                                    appctx, options_prefix,
-                                                    zero_bc_nodes, weight,
-                                                    *(visited[arg] for arg in operands))
     if tensor:
-        update_tensor(visited[expr], tensor)
-    return visited[expr]
+        update_tensor(result, tensor)
+    return result
+
+
+def preprocess_base_form(expr, mat_type=None, form_compiler_parameters=None):
+    """Preprocess ufl.BaseForm objects"""
+    if mat_type != "matfree":
+        # For "matfree", Form evaluation is delayed
+        expr = expand_derivatives_form(expr, form_compiler_parameters)
+    # Expanding derivatives may turn `ufl.BaseForm` objects into `ufl.Expr` objects that are not `ufl.BaseForm`.
+    if not isinstance(expr, ufl.form.BaseForm):
+        return assemble(expr)
```

Comment:
It strikes me as a bit odd to call `assemble` inside of a preprocessing function. It seems like it should recurse infinitely.

Reply:
No, because the object is not a `ufl.form.BaseForm` anymore. Calling back `assemble` to use the right assembly procedure at this stage is a consequence of expanding derivatives (cf. above comment).

Reply:
My point is that isn't obviously preprocessing. The name of the function is a bit misleading.

Reply:
The point of this function is really to preprocess the form via expanding derivatives or restructuring the corresponding DAG. I think what I have done here might be a premature simplification. I should be able to lift that out of the preprocessing stage so that it's done within a function aimed at assembling and not preprocessing, which would make things clearer.

Reply:
Done